aerogear/app-metrics-operator
# Add backups to AppMetricsService CRD & controller

JIRA: https://issues.jboss.org/browse/AEROGEAR-9571

# Verification steps

* Check out this PR
* Deploy the operator
```
make cluster/prepare
sed -e "s|quay.io/aerogear/app-metrics-operator:master|quay.io/aerogear/app-metrics-operator:$(git rev-parse --short HEAD)|g" deploy/operator.yaml | kubectl apply -n app-metrics -f -
```
* Create a ServiceAccount called `backupjob`. This is needed for the resulting cronjobs, but is not created by the operator, since it may need additional access that the operator may not be able to give -- if the secrets being used are in different namespaces, then there would need to be relevant rolebindings for it, so we deliberately leave out the creation of all of that stuff:
```
kubectl create sa backupjob

# Need to add this so that it can read secrets in the app-metrics namespace. If using secrets in a different namespaces, run this command for those namespaces too.
oc policy add-role-to-user edit system:serviceaccount:app-metrics:backupjob -n app-metrics
```
* Create an AppMetricsService CR that has backups specified (if you plan to test the backups, you'll need to modify this to put the correct secrets' coordinates in place:
```
kubectl apply -f deploy/crds/metrics_v1alpha1_appmetricsservice_cr_with_backups.yaml
```

* Verify that there are no persistent errors in the operator logs, and that everything gets created as expected. The important things to check for here are "cronjob" objects:

```
kubectl get cronjob -l app=example-appmetricsservice-with-backups
NAME                           SCHEDULE    SUSPEND   ACTIVE    LAST SCHEDULE   AGE
appmetrics-daily-at-midnight   0 0 * * *   False     0         <none>          1h
```

* Edit the schedule of the cronjob, and notice that it gets changed back immediately (note the output may not look exactly like this, due to concurrency, I've annotated with comments):
```
$ oc get cronjobs -l app=example-appmetricsservice-with-backups -w & oc patch cronjob appmetrics-daily-at-midnight -p '{"spec": {"schedule": "1 0 * * *"}}'
[1] 17231 #pid of backgrounded `oc get -w` task, you can kill it after this step, e.g. `kill 17231`
NAME                    SCHEDULE    SUSPEND   ACTIVE    LAST SCHEDULE   AGE
ups-daily-at-midnight   0 0 * * *   False     0         <none>          4m #before patch, as specified in CR
cronjob.batch/ups-daily-at-midnight patched
ups-daily-at-midnight   1 0 * * *   False     0         <none>    4m #after patch
ups-daily-at-midnight   0 0 * * *   False     0         <none>    4m #after reconcile fixes it back to what the CR specified
```
